### PR TITLE
rust: cargo: prefer primary spans

### PR DESCRIPTION
**Pull request by dpc**
This would be great improvement to the current situation, where the
first span is used, which is quite often not the right one.

Note: My VimL is poor, but I'm desperate to get this fixed.

See https://github.com/rust-lang/rust.vim/issues/174

Related to: #2075

### CODE REVIEW

File: autoload/neomake/makers/ft/rust.vim [hunk 1]
```diff
@@ -64,6 +64,12 @@ function! neomake#makers#ft#rust#CargoProcessOutput(context) abort
         endif
 
         let span = data.spans[0]
+        for candidate_span in data.spans
+            if candidate_span.is_primary
+                let span = candidate_span
```
Comment (blueyed) [inline autoload/neomake/makers/ft/rust.vim]: `break`?

**Comment by blueyed** [comment]
Would also be good to have a test for this.
There are tests already, where likely only some fixture would need to be adjusted to fail before, but succeed afterwards.
/cc @fuine

File: autoload/neomake/makers/ft/rust.vim (hunk 1 above)
Comment (blueyed) [reply autoload/neomake/makers/ft/rust.vim]: According to https://github.com/rust-lang/rust.vim/issues/174#issuecomment-419472174 the first primary span should be used.

**Comment by dpc** [comment]
I've added the `break`, but I've gave up on tests, as I can't figure it all out.


To help anyone willing to help me, this is a simple project that can be used as fixture. The error should point to line 6, not line 1.

```
> cat src/main.rs
fn foo(i : i32) {
    unimplemented!();
}

fn bar() {
    foo(1, 1);
}

fn main() {
    println!("Hello, world!");
}


> cargo build --message-format=json
   Compiling testfoo v0.1.0 (file:///home/dpc/tmp/testfoo)
{"message":{"children":[],"code":{"code":"E0061","explanation":"\nThe number of arguments passed to a function must match the number of arguments\nspecified in the function signature.\n\nFor example, a function like:\n\n```\nfn f(a: u16, b: &str) {}\n```\n\nMust always be called with exactly two arguments, e.g. `f(2, \"test\")`.\n\nNote that Rust does not have a notion of optional function arguments or\nvariadic functions (except for its C-FFI).\n"},"level":"error","message":"this function takes 1 parameter but 2 parameters were supplied","rendered":"error[E0061]: this function takes 1 parameter but 2 parameters were supplied\n --> src/main.rs:6:5\n  |\n1 | fn foo(i : i32) {\n  | --------------- defined here\n...\n6 |     foo(1, 1);\n  |     ^^^^^^^^^ expected 1 parameter\n\n","spans":[{"byte_end":15,"byte_start":0,"column_end":16,"column_start":1,"expansion":null,"file_name":"src/main.rs","is_primary":false,"label":"defined here","line_end":1,"line_start":1,"suggested_replacement":null,"suggestion_applicability":null,"text":[{"highlight_end":16,"highlight_start":1,"text":"fn foo(i : i32) {"}]},{"byte_end":67,"byte_start":58,"column_end":14,"column_start":5,"expansion":null,"file_name":"src/main.rs","is_primary":true,"label":"expected 1 parameter","line_end":6,"line_start":6,"suggested_replacement":null,"suggestion_applicability":null,"text":[{"highlight_end":14,"highlight_start":5,"text":"    foo(1, 1);"}]}]},"package_id":"testfoo 0.1.0 (path+file:///home/dpc/tmp/testfoo)","reason":"compiler-message","target":{"crate_types":["bin"],"edition":"2015","kind":["bin"],"name":"testfoo","src_path":"/home/dpc/tmp/testfoo/src/main.rs"}}
{"message":{"children":[],"code":null,"level":"error","message":"aborting due to previous error","rendered":"error: aborting due to previous error\n\n","spans":[]},"package_id":"testfoo 0.1.0 (path+file:///home/dpc/tmp/testfoo)","reason":"compiler-message","target":{"crate_types":["bin"],"edition":"2015","kind":["bin"],"name":"testfoo","src_path":"/home/dpc/tmp/testfoo/src/main.rs"}}
{"message":{"children":[],"code":null,"level":"","message":"For more information about this error, try `rustc --explain E0061`.","rendered":"For more information about this error, try `rustc --explain E0061`.\n","spans":[]},"package_id":"testfoo 0.1.0 (path+file:///home/dpc/tmp/testfoo)","reason":"compiler-message","target":{"crate_types":["bin"],"edition":"2015","kind":["bin"],"name":"testfoo","src_path":"/home/dpc/tmp/testfoo/src/main.rs"}}
error: Could not compile `testfoo`.
```

**Comment by dpc** [comment]
```
{"message":{"children":[],"code":{"code":"E0061","explanation":"\nThe number of arguments passed to a function must match the number of arguments\nspecified in the function signature.\n\nFor example, a function like:\n\n```\nfn f(a: u16, b: &str) {}\n```\n\nMust always be called with exactly two arguments, e.g. `f(2, \"test\")`.\n\nNote that Rust does not have a notion of optional function arguments or\nvariadic functions (except for its C-FFI).\n"},"level":"error","message":"this function takes 1 parameter but 2 parameters were supplied","rendered":"error[E0061]: this function takes 1 parameter but 2 parameters were supplied\n --> src/main.rs:6:5\n  |\n1 | fn foo(i : i32) {\n  | --------------- defined here\n...\n6 |     foo(1, 1);\n  |     ^^^^^^^^^ expected 1 parameter\n\n","spans":[{"byte_end":15,"byte_start":0,"column_end":16,"column_start":1,"expansion":null,"file_name":"src/main.rs","is_primary":false,"label":"defined here","line_end":1,"line_start":1,"suggested_replacement":null,"suggestion_applicability":null,"text":[{"highlight_end":16,"highlight_start":1,"text":"fn foo(i : i32) {"}]},{"byte_end":67,"byte_start":58,"column_end":14,"column_start":5,"expansion":null,"file_name":"src/main.rs","is_primary":true,"label":"expected 1 parameter","line_end":6,"line_start":6,"suggested_replacement":null,"suggestion_applicability":null,"text":[{"highlight_end":14,"highlight_start":5,"text":"    foo(1, 1);"}]}]},"package_id":"testfoo 0.1.0 (path+file:///home/dpc/tmp/testfoo)","reason":"compiler-message","target":{"crate_types":["bin"],"edition":"2015","kind":["bin"],"name":"testfoo","src_path":"/home/dpc/tmp/testfoo/src/main.rs"}}
```

**Comment by blueyed** [comment]
Working on this now.
Relevant: https://github.com/neomake/neomake/pull/2081 (without this the `filename` might not get set properly on entries - likely a main reason/confusion for the cwd issues you are seeing).

**Comment by blueyed** [comment]
Rebased and amended with a test.

**Comment by blueyed** [comment]
Thanks!

**Pull request merged**